---
title: "A *plyranges* and *tximeta* case study"
subtitle: "Overlapping differential expression and chromatin accessibility"
author: "Michael Love, Stuart Lee, Michael Lawrence"
---

```{r setup, echo=FALSE, results="hide"}
knitr::opts_chunk$set(cache=TRUE)
```

# RNA-seq data import

```{r}
suppressPackageStartupMessages(library(plyranges))
library(macrophage)
dir <- system.file("extdata", package="macrophage")
coldata <- file.path(dir, "coldata.csv") %>% 
  read.csv() %>% 
  dplyr::select(
    names, 
    id = sample_id, 
    line = line_id, 
    condition = condition_name
  ) %>%
  mutate(
    files = file.path(dir, "quants", names, "quant.sf.gz"),
    condition = relevel(condition, "naive")
  )
head(coldata)
```

```{r}
suppressPackageStartupMessages(library(SummarizedExperiment))
library(tximeta)
```

```{r}
se <- tximeta(coldata, dropInfReps=TRUE)
assayNames(se)
```

```{r}
gse <- summarizeToGene(se)
```

# Basic DE task

```{r}
library(DESeq2)
dds <- DESeqDataSet(gse, ~line + condition)
keep <- rowSums(counts(dds) >= 10) >= 6
dds <- dds[keep,]
```

```{r}
dds <- DESeq(dds)
res <- results(dds, contrast=c("condition","IFNg","naive"),
               lfcThreshold=1)
```

Always good to do a little visualization of results:

```{r}
summary(res, alpha=.01)
DESeq2::plotMA(res, ylim=c(-10,10))
plotCounts(dds, which.min(res$pvalue), "condition")
```

Output *GRanges* results (both DE and some non-DE for comparison
later): 

```{r}
de_genes <- results(dds, 
                    contrast=c("condition","IFNg","naive"),
                    lfcThreshold=1, 
                    format="GRanges") %>%
  filter(padj < 0.01) %>%
  mutate(gene_id=names(.)) %>%
  plyranges::select(gene_id,
                    de_log2FC = log2FoldChange,
                    de_padj = padj)
```

We re-run `results` because we don't want to use an `lfcThreshold` 
this time:

```{r}
other_genes <- results(dds, 
                       contrast=c("condition","IFNg","naive"),
                       format="GRanges") %>% 
  filter(pvalue > 0.1) %>%
  mutate(gene_id=names(.)) %>%
  plyranges::select(gene_id,
                    de_log2FC = log2FoldChange,
                    de_padj = padj)
```

# ATAC peak data

<https://zenodo.org/record/1188300#.XIAhXlNKjOQ>

This takes ~30 seconds and loads an object of ~370 Mb.

```{r}
atac_mat <- as.matrix(read.delim("ATAC_cqn_matrix.txt.gz"))
```

```{r}
atac_coldata <- read.delim("ATAC_sample_metadata.txt.gz") %>% 
  plyranges::select(
    sample_id,
    donor,
    condition = condition_name
  ) %>% 
  mutate(condition = relevel(condition, "naive"))
```

```{r}
peaks_df <- read.delim("ATAC_peak_metadata.txt.gz", strings=FALSE)
peaks_gr <- peaks_df %>% 
  plyranges::select(seqnames = chr, start, end, peak_id=gene_id) %>% 
  as_granges()
# we know this from the Zenodo entry
# https://zenodo.org/record/1188300#.XJOFSlNKiL5
genome(peaks_gr) <- "GRCh38"
```

```{r}
idx <- match(colnames(atac_mat), atac_coldata$sample_id)
atac_coldata <- atac_coldata[idx,]
all.equal(colnames(atac_mat), as.character(atac_coldata$sample_id))
```

```{r}
atac <- SummarizedExperiment(list(cqndata=atac_mat),
                             rowRanges=peaks_gr,
                             colData=atac_coldata)
```

Check the SD over mean plot:

```{r}
rmu <- rowMeans(assay(atac))
rvar <- rowVars(assay(atac))
idx <- sample(nrow(atac),1e4)
plot(rmu[idx], sqrt(rvar[idx]), cex=.1)
```

```{r}
library(limma)
design <- model.matrix(~donor + condition, colData(atac))
fit <- lmFit(assay(atac), design)
fit <- eBayes(fit)
idx <- which(colnames(fit$coefficients) == "conditionIFNg")
tt <- topTable(fit, coef=idx, sort.by="none", n=nrow(atac))
```

Plot the top gene by LFC

```{r}
idx <- which.max(tt$logFC)
plot(assay(atac)[idx,] ~ atac$condition)
table(tt$logFC > 1 & tt$adj.P.Val < .05)
```

```{r}
peaks <- rowRanges(atac) %>% 
  mutate(
    da_log2FC = tt$logFC,
    da_padj = tt$adj.P.Val
  )
seqlevelsStyle(peaks) <- "UCSC"
genome(peaks) <- "hg38"
```

```{r}
da_peaks <- peaks %>% filter(da_padj < .01)
```

# Overlap analysis

```{r}
# sub-sample, here define a dplyr function first
sample_n <- function(x,n) x[sample.int(length(x),n,replace=FALSE)]
```

```{r}
other_genes <- other_genes %>%
  sample_n(length(de_genes))
```

```{r}
lvls <- c("not_de","de")
all_genes <- bind_ranges(de=de_genes,
                         not_de=other_genes,
                         .id="origin") %>%
  mutate(
    origin = factor(sub("de.*", "de", origin), lvls)
  )
all_genes
``` 

```{r}
overlap_genes <- all_genes %>%
  mutate(width=1e4) %>%
  join_overlap_left(da_peaks)
overlap_genes
```

How many DA peaks are near DE genes (where a gene may appear more than
once, because we are doing a join):

```{r}
overlap_tab <- overlap_genes %>%
  group_by(origin) %>%
  summarize(any=sum(!is.na(da_padj)),
            lfc1=sum(abs(da_log2FC) > 1, na.rm=TRUE),
            lfc2=sum(abs(da_log2FC) > 2, na.rm=TRUE))
overlap_tab
```

```{r}
sapply(overlap_tab[,-1], function(x) x[2]/x[1])
```

We can ask for the top peak per gene based on the peak's LFC:

```{r}
library(ggplot2)
overlap_genes %>%
  group_by(gene_id) %>%
  plyranges::filter(abs(da_log2FC) == max(abs(da_log2FC))) %>%
  mcols %>% as.data.frame %>%
  ggplot(aes(origin, abs(da_log2FC))) +
  geom_boxplot()
```
